### ISSUE TITLE
Sort-Object -Stable -Unique is not always stable

### ISSUE THREAD

**Issue by m1k0net** [PowerShell/PowerShell#16907]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

When running data through Sort-Object -Stable -Unique given a specific property the returned results do not appear to be the 1st element seen from a stable sort.  Also modifying modifying the input data for an unrelated record effects the records of another.  Below is some sample code that can be used to reproduce the issue.  The only difference from the data in the 1st set to the 2nd is the last row was changed from 2021 to 2015 which is effecting the file1 key but the file0 key is consistently showing the last item.  The rest of the records appear to process correctly.

```powershell
$Data="Key,Path,Size
file1,C:\folder\2015,10
file1,C:\folder\2018,10
file1,C:\folder\2020,10
file2,C:\folder\2014,10
file2,C:\folder\2019,10
file4,C:\folder\2020,10
file4,C:\folder\2017,10
file4,C:\folder\2020,10
file4,C:\folder\2019,10
file4,C:\folder\2021,10
file0,C:\folder\2015,10
file0,C:\folder\2018,10
file0,C:\folder\2020,10
file2,C:\folder\2022,10
file3,C:\folder\2019,10
file3,C:\folder\2018,10
file3,C:\folder\2019,10
file3,C:\folder\2020,10
file3,C:\folder\2022,10
file3,C:\folder\2021,10"
$List=$Data | ConvertFrom-Csv  
$SortedList=$List | Sort-Object Path -Descending
Write-Host "Sorted List:"
$SortedList | Format-Table
Write-Host "Unique List:"
$SortedList | Sort-Object Key -Unique -Stable | Format-Table


$list[19].Path="C:\folder\2015"
$SortedList=$List | Sort-Object Path -Descending
Write-Host "Sorted List:"
$SortedList | Format-Table
Write-Host "Unique List:"
$SortedList | Sort-Object Key -Unique -Stable | Format-Table```


### Expected behavior

```console
Key   Path           Size
---   ----           ----
file0 C:\folder\2020 10
file1 C:\folder\2020 10
file2 C:\folder\2022 10
file3 C:\folder\2022 10
file4 C:\folder\2021 10
```

### Actual behavior

```console
Case 1:
Key   Path           Size
---   ----           ----
file0 C:\folder\2015 10
file1 C:\folder\2015 10
file2 C:\folder\2022 10
file3 C:\folder\2022 10
file4 C:\folder\2021 10

Case 2:
Key   Path           Size
---   ----           ----
file0 C:\folder\2015 10
file1 C:\folder\2020 10
file2 C:\folder\2022 10
file3 C:\folder\2022 10
file4 C:\folder\2021 10
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.2.1
PSEdition                      Core
GitCommitId                    7.2.1
OS                             Microsoft Windows 10.0.19043
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0


Name                           Value
----                           -----
PSVersion                      7.3.0-preview.1
PSEdition                      Core
GitCommitId                    7.3.0-preview.1
OS                             Microsoft Windows 10.0.19043
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by m1k0net** [comment]
I believe I found the issue.  In Sort-Object.cs when a duplicate is found an item is taken from the end of the array and moved to the current position.  This has the side effect of breaking stable sorts because the source array is being re-ordered.  Below is the relevant code.  I believe a fix would just be instead of re-ordering the array just skip to the next record.  It seems that would be quicker and save on the data copying but I don't know enough C# to know if that is proper.

```c#
                   if (dataIndex != dataToSort.Count - discardedDuplicates)
                    {
                        // When discarding duplicates, replace them with an item at the end of the list and
                        // adjust our counter so that we check the item we just swapped in next
                        dataToSort[dataIndex] = dataToSort[dataToSort.Count - discardedDuplicates];
                        dataIndex--;
                    }
```

I have created the below targeted test case for the issue.  Currently the test fails since the Record "B2" is returned instead of "B1".

```powershell
    It "Sort-Object with Uniqe Stable should return proper records"{
        $Record1=[PSCustomObject]@{Key="A";Record="A1"}
        $Record2=[PSCustomObject]@{Key="A";Record="A2"}
        $Record3=[PSCustomObject]@{Key="B";Record="B1"}
        $Record4=[PSCustomObject]@{Key="B";Record="B2"}
        $Records = @($Record1,$Record2,$Record3,$Record4)
        $results = $Records | Sort-Object -Stable -Unique -Property Key

        $results[0] | Should -Be $Records[0]
        $results[1] | Should -Be $Records[2]
    }
```

**Comment by m1k0net** [comment]
This was resolved in PR #17189
